Core/Src/main.c

```diff
@@ -1,181 +1,253 @@
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+uint8_t adc_start_byte = 1;
+uint8_t adc_config_byte = 0;
+uint8_t adc_rec_buf = 0;
+
+uint16_t adc_val = 0;
+double percentage_duty_cycle = 0.0f;
+const uint16_t MAX_VAL_IN_TEN_BITS = 1023;
+
+const uint16_t COUNT_PER_MS = 64000 / 20;
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+uint8_t setupAdcConfigByte();
+void copyBits(uint16_t* to, uint8_t from, uint8_t num_bits);
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+  adc_config_byte = setupAdcConfigByte();
+  // Set CS pin to high initially
+  if (HAL_GPIO_ReadPin(GPIOB, GPIO_PIN_8)) {
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+  }
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+    /* Grab ADC value */
+    // Set CS to low
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+
+    // Send start bit
+    HAL_SPI_Transmit(&hspi1, &adc_start_byte, 1, 100);
+
+    // Send ADC configs, receive first 3 bits of data
+    HAL_SPI_TransmitReceive(&hspi1, &adc_config_byte, &adc_rec_buf, 1, 100);
+    copyBits(&adc_val, adc_rec_buf, 2);
+
+    // Receive next 8 bits of data
+    HAL_SPI_Receive(&hspi1, &adc_rec_buf, 1, 100);
+    copyBits(&adc_val, adc_rec_buf, 8);

```

Comment:
I see what you are trying to do here but in the end you are only receiving a total of 2 bytes (the first two receive bytes in the figure below), the first byte received is from your `HAL_SPI_TransmitReceive` function and the second byte is from your `HAL_SPI_Receive` function. This means you are missing the last byte from the image below, B7 - B0.

What you can do instead is use only the `HAL_SPI_TransmitReceive` function and send and receive 3 bytes. You can accomplish this by populating a "send" array of size 3 with the 3 bytes of data you want to send (the last byte is all DON'T CARE, as indicated by the figure below, which means you can send anything you want, it doesn't matter) and then receiving in a "receive" array of size 3. Then after, all you have to do is stitch the receive data together to get your raw adc value.

If this doesn't make sense please let me know!

![image](https://user-images.githubusercontent.com/69175242/230981416-0ee2447f-4931-4b80-a1c6-e036a23a7964.png)